src/common/utils/ldap/ldap.go

```diff
@@ -0,0 +1,367 @@
+		filterTag = "*"
+	} else {
+		filterTag = username
+	}
+
+	if ldapFilter == "" {
+		ldapFilter = "(" + ldapUID + "=" + filterTag + ")"
+	} else {
+		if !strings.Contains(ldapFilter, ldapUID+"=") {
+			ldapFilter = "(&" + ldapFilter + "(" + ldapUID + "=" + filterTag + "))"
+		} else {
+			ldapFilter = strings.Replace(ldapFilter, ldapUID+"=*", ldapUID+"="+username, -1)
+		}
+	}
+
+	log.Debug("one or more ldapFilter: ", ldapFilter)
+
+	return ldapFilter
+}
+
+// ConnectTest ...
+func ConnectTest(ldapConfs models.LdapConf) error {
+
+	var ldapConn *goldap.Conn
+	var err error
+
+	ldapConn, err = dailLDAP(ldapConfs, ldapConn)
+
+	if err != nil {
+		return err
+	}
+	defer ldapConn.Close()
+
+	if ldapConfs.LdapSearchDn != "" {
+		err = bindLDAP(ldapConfs, ldapConn)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// SearchUser ...
+func SearchUser(ldapConfs models.LdapConf) ([]models.LdapUser, error) {
+	var ldapUsers []models.LdapUser
+	var ldapConn *goldap.Conn
+	var err error
+
+	ldapConn, err = dailLDAP(ldapConfs, ldapConn)
+
+	if err != nil {
+		return nil, err
+	}
+	defer ldapConn.Close()
+
+	if ldapConfs.LdapSearchDn != "" {
+		err = bindLDAP(ldapConfs, ldapConn)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if ldapConfs.LdapBaseDn == "" {
+		return nil, fmt.Errorf("can not get any available LDAP_BASE_DN")
+	}
+
+	result, err := searchLDAP(ldapConfs, ldapConn)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ldapEntry := range result.Entries {
+		var u models.LdapUser
+		for _, attr := range ldapEntry.Attributes {
+			val := attr.Values[0]
+			switch attr.Name {
+			case ldapConfs.LdapUID:
+				u.Username = val
+			case "uid":
+				u.Realname = val
+			case "cn":
+				u.Realname = val
+			case "mail":
+				u.Email = val
+			case "email":
+				u.Email = val
+			}
+		}
+		ldapUsers = append(ldapUsers, u)
+	}
+
+	return ldapUsers, nil
+}
+
+func formatLdapURL(ldapURL string) (string, error) {
+
+	var protocol, hostport string
+	var err error
+
+	_, err = url.Parse(ldapURL)
+	if err != nil {
+		return "", fmt.Errorf("parse Ldap Host ERR: %s", err)
+	}
+
+	if strings.Contains(ldapURL, "://") {
+		splitLdapURL := strings.Split(ldapURL, "://")
+		protocol, hostport = splitLdapURL[0], splitLdapURL[1]
+		if !((protocol == "ldap") || (protocol == "ldaps")) {
+			return "", fmt.Errorf("unknown ldap protocl")
+		}
+	} else {
+		hostport = ldapURL
+		protocol = "ldap"
+	}
+
+	if strings.Contains(hostport, ":") {
+		splitHostPort := strings.Split(hostport, ":")
+		port, error := strconv.Atoi(splitHostPort[1])
+		if error != nil {
+			return "", fmt.Errorf("illegal url port")
+		}
+		if port == 636 {
+			protocol = "ldaps"
+		}
+
+	} else {
+		switch protocol {
+		case "ldap":
+			hostport = hostport + ":389"
+		case "ldaps":
+			hostport = hostport + ":636"
+		}
+	}
+
+	fLdapURL := protocol + "://" + hostport
+
+	return fLdapURL, nil
+
+}
+
+// ImportUser ...
+func ImportUser(user models.LdapUser) (int64, string) {
```

Comment:
because this feature will return a long array. 
`type LdapFailedImportUser struct {`
`	UID   string `json:"uid"` `
`	Error string `json:"err_msg"` `
`}`
UI want string for easy process